opm/autodiff/ExtendedBlackoilModel.hpp

```diff
@@ -0,0 +1,286 @@
+                             const NewtonIterationBlackoilInterface& linsolver,
+                             const EclipseStateConstPtr              eclState,
+                             const bool                              has_disgas,
+                             const bool                              has_vapoil,
+                             const bool                              terminal_output,
+                             const bool                              has_solvent);
+
+        /// Called once before each time step.
+        /// \param[in] dt                     time step size
+        /// \param[in, out] reservoir_state   reservoir state variables
+        /// \param[in, out] well_state        well state variables
+        void prepareStep(const double dt,
+                         ReservoirState& reservoir_state,
+                         WellState& well_state);
+
+        /// Called once after each time step.
+        /// \param[in] dt                     time step size
+        /// \param[in, out] reservoir_state   reservoir state variables
+        /// \param[in, out] well_state        well state variables
+        void afterStep(const double dt,
+                       ReservoirState& reservoir_state,
+                       WellState& well_state);
+
+        /// Apply an update to the primary variables, chopped if appropriate.
+        /// \param[in]      dx                updates to apply to primary variables
+        /// \param[in, out] reservoir_state   reservoir state variables
+        /// \param[in, out] well_state        well state variables
+        void updateState(const V& dx,
+                         ReservoirState& reservoir_state,
+                         WellState& well_state);
+
+        /// Compute convergence based on total mass balance (tol_mb) and maximum
+        /// residual mass balance (tol_cnv).
+        /// \param[in]   dt          timestep length
+        /// \param[in]   iteration   current iteration number
+        bool getConvergence(const double dt, const int iteration);
+
+        /// Assemble the residual and Jacobian of the nonlinear system.
+        /// \param[in]      reservoir_state   reservoir state variables
+        /// \param[in, out] well_state        well state variables
+        /// \param[in]      initial_assembly  pass true if this is the first call to assemble() in this timestep
+        void assemble(const ReservoirState& reservoir_state,
+                      WellState& well_state,
+                      const bool initial_assembly);
+
+
+    protected:
+
+        // ---------  Types and enums  ---------
+
+        typedef typename Base::SolutionState SolutionState;
+        typedef typename Base::DataBlock DataBlock;
+        enum { Solvent = CanonicalVariablePositions::Next };
+
+        // ---------  Data members  ---------
+        const bool has_solvent_;
+        const int solvent_pos_;
+        const SolventPropsAdFromDeck solvent_props_;
```

Comment:
This should be a reference to be consistent with the polymer solver (and avoid copying). We may consider changing this after refactoring, but for now it should be a reference.